Sources/Amplitude/Amplitude.m

```diff
@@ -483,6 +483,7 @@ - (void) checkInForeground {
     // notification is already triggered, so we need to manually check and set it now.
     // UIApplication methods are only allowed on the main thread so need to dispatch this synchronously to the main thread.
     void (^checkInForeground)(void) = ^{
+        self->_inForeground = YES;
```

Comment:
If we set the value here we are not really doing any check right?

Reply:
I thought the following ensured that this call is run in the foreground. Is that not the case?
`[self runSynchronouslyOnMainQueue:checkInForeground];`

Reply:
@qingzhuozhen do you mean comment ` we need to manually check`?
I thought:
* when application starts: `checkInForegroung` logic works, `enterForeground` doesn't
* when application goes to foreground: `enterForegroung` logic works, `checkInForeground` doesn't
* `enterForegroung` and `checkInForeground`  do not work together
* _inForeground should be `NO` at start of both `enterForegroung` (`NO` should be set by previous `enterBackground`) and `checkInForeground` (`NO` is initial value).
* `checkInForeground` doesn't check anything and can be renamed

Does it look correct?

Reply:
I can change `checkInForeground` assuming it can work only on app initialization:
```
void (^checkInForeground)(void) = ^{
    BOOL inForeground = _inForeground;
    self->_inForeground = YES;
    ...
                [self startOrContinueSessionNSNumber:now inForeground:inForeground];
    ...            
```

Reply:
I guess the purpose of `if (state != UIApplicationStateBackground)` is to check if we are on foreground or background? 
From the comment, I guess the issue could be when initialized the SDK, it could be already on foreground state, and since the observer wasn't set earlier, enterForeground is possible for not being called. That's probably why we need this checkInForeground method. Another possible scenario is if initialize happens in background, and log a event to refresh the session. And after init, the foreground is mostly relying on the enterForeground. The assumptions are basically right though I guess the checkInForeground method still try to a state check there.

Reply:
@qingzhuozhen Thank you, now it is clearer for me. Let me think a bit.

Reply:
@qingzhuozhen Now `self->_inForeground = YES;` setter is in `if (state != UIApplicationStateBackground)` body